syntax = "proto3";
package types;

option go_package = "github.com/canopy-network/go-plugin/contract";

import "google/protobuf/any.proto";

message Event {
  // event_type: The type of the transaction like 'slash' or 'unstake'
  string event_type = 1; // @gotags: json:"eventType"
  // msg: The actual event message payload, which is encapsulated in a oneof format
  oneof msg {
    EventCustom custom = 11;
  }
  // height: the block height of the event
  uint64 height = 91;
  // reference: the 'begin_block' / transaction hash / 'end_block'
  string reference = 92;
  // chainId: the chain id associated with the event
  uint64 chainId = 93;
  // block_height: the number of blocks in the blockchain before a specific block, indicating its position in the chain
  // and representing the blockchain's length
  uint64 block_height = 94;  // @gotags: json:"blockHeight"
  // block_hash: a unique identifier for a block, generated by applying a cryptographic hash function to the block's contents,
  // ensuring its integrity and connecting it to the previous block in the chain.
  bytes block_hash = 95;  // @gotags: json:"blockHash"
  // address: The address of the user affected by the event
  bytes address = 96;
}

// EventCustom carries a plugin-defined event payload.
message EventCustom {
  // msg: custom payload.
  google.protobuf.Any msg = 1;
}
